src/modules/aws_ec2/mappers/instance_block_device_mapping.ts

```diff
@@ -0,0 +1,282 @@
+  };
+
+  waitUntilInUse(client: EC2, volumeId: string) {
+    return this.module.generalPurposeVolume.volumeWaiter(client, volumeId, (vol: AWSVolume | undefined) => {
+      // If state is not 'in-use' retry
+      if (!Object.is(vol?.State, VolumeState.IN_USE)) {
+        return { state: WaiterState.RETRY };
+      }
+      return { state: WaiterState.SUCCESS };
+    });
+  }
+
+  attachVolume = async (client: EC2, VolumeId: string, InstanceId: string, Device: string) => {
+    const result = await this.attachVolumeInternal(client, VolumeId, InstanceId, Device);
+    if (result) await this.waitUntilInUse(client, VolumeId);
+    else return false;
+    return true;
+  };
+
+  cloud: Crud2<InstanceBlockDeviceMapping> = new Crud2({
+    create: async (es: InstanceBlockDeviceMapping[], ctx: Context) => {
+      const out: InstanceBlockDeviceMapping[] = [];
+      for (const e of es) {
+        // if we have different regions we may be in a multi-region update situation, just skip
+        if (e.region !== e.instance.region) continue;
+
+        // if instance is not created we are in the first step, no need to create anything
+        if (!e.instance?.instanceId) continue;
+        const client = (await ctx.getAwsClient(e.region)) as AWS;
+
+        // read volume details
+        if (!e.volumeId) throw new Error('Cannot attach empty volumes to an instance already created');
+        if (!e.volume?.volumeId) throw new Error('Tried to attach an unexisting volume');
+        if (e.volume.region !== e.instance.region)
+          throw new Error('Cannot create a mapping between different regions');
+
+        // only attach if volume is not in use. We can have the case that the volume is auto-attached on creation
+        if (e.volume.state === VolumeState.AVAILABLE) {
+          await this.attachVolume(client.ec2client, e.volume.volumeId, e.instance.instanceId, e.deviceName);
+        }
+
+        // return mapping
+        const newMap: InstanceBlockDeviceMapping = {
+          deviceName: e.deviceName,
+          instanceId: e.instanceId,
+          instance: e.instance,
+          volumeId: e.volumeId,
+          volume: e.volume,
+          region: e.instance.region,
+          deleteOnTermination: e.deleteOnTermination,
+        };
+
+        // Save the record back into the database to get the new fields updated
+        newMap.id = e.id;
+        await this.module.instanceBlockDeviceMapping.db.update(newMap, ctx);
+        out.push(newMap);
+      }
+      return out;
+    },
+    read: async (ctx: Context, id?: string) => {
+      if (id) {
+        // decompose the id
+        const { instanceId, volumeId, region } = this.idFields(id);
+        const client = (await ctx.getAwsClient(region)) as AWS;
+
+        // check if we can find the instance in database
+        const instance = await this.module.instance.db.read(
+          ctx,
+          this.module.instance.generateId({ instanceId: instanceId ?? '', region }),
+        );
```

Comment:
Since this is the `cloud.read` path, perhaps it should call `this.module.instance.cloud.read` first and if that fails it can fall back to the `db.read` path?

Reply:
i did it because this mapper actually needs to have the records existing in the database, see the mapper as 
```      const res: InstanceBlockDeviceMapping = {
        instanceId: instance?.id ?? undefined,
        volumeId: volume?.id ?? undefined,
        instance,
        volume,
        deviceName: newMap.DeviceName,
        region,
        deleteOnTermination: newMap.Ebs?.DeleteOnTermination ?? true,
      };
```
so i thought that it would not make sense to query for records that are not already created